File: Data Structures & Algorithms/k-closest-points-to-origin/submission-2.js
class Solution {
    /**
     * @param {number[][]} points
     * @param {number} k
     * @return {number[][]}
     */
    quickSort(arr, k, s = 0, e = arr.length - 1) {
        if (e - s + 1 <= 1) return arr;

        let pivot = arr[e];
        let left = s;

        for (let i = s; i < e; i++) {
            const distI = arr[i][0] ** 2 + arr[i][1] ** 2;
            const distpivot = pivot[0] ** 2 + pivot[1] ** 2;

            if (distI < distpivot) {
                let temp = arr[left];
                arr[left] = arr[i];
                arr[i] = temp;
                left++;
            }
        }

        arr[e] = arr[left];
        arr[left] = pivot;

        if (left === k) {
            return arr.slice(0, k);
        }
        if (left < k) {
            this.quickSort(arr, k, left + 1, e);
        } else {
            this.quickSort(arr, k, s, left - 1);
        }

        return arr.slice(0, k);
    }
    kClosest(points, k) {
        return this.quickSort(points, k);
    }
}

File: Data Structures & Algorithms/quickSort/submission-2.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def quickSort(self, pairs: List[Pair], s: int = 0, e: Optional[int] = None) -> List[Pair]:
        if e == None:
            e = len(pairs) - 1

        if e - s + 1 <= 1:
            return pairs

        pivot = pairs[e]
        left = s
        for i in range(s, e):
            if pairs[i].key < pivot.key:
                pairs[left], pairs[i] = pairs[i], pairs[left]
                left += 1

        pairs[e] = pairs[left]
        pairs[left] = pivot

        self.quickSort(pairs, s, left - 1)
        self.quickSort(pairs, left + 1, e)

        return pairs
